KEP-4330: Change API availability to forward compatibility (#5038)
kubernetes/enhancements: keps/sig-architecture/4330-compatibility-versions/README.md

Author: sometimes if you know when the api would be removed you can just add the removal before the actual release.

```

#### Un-Emulatable Features
In very rare cases, if the feature implementation history could not be preserved in the code base with reasonable maintenance cost, a feature could opt out of the compatibility version through the `UnEmulatableFeatures` list. 

With the `UnEmulatableFeatures` list, the server would fail to start if
1. the `EmulationVersion != BinaryVersion` and
1. any feature in the `UnEmulatableFeatures` list is disabled by default at the emulation version and explicitly enabled by the `--feature-gates` flag.

So this exception should only be used if the feature  
1. is Beta and disabled by default for at least one version in the range of `1.{binaryMinorVersion-3}..1.{binaryMinorVersion-1}`.
1. is going through a lot of changes when it is off by default, and the feature implementation history is very hard to preserve.

> 1. has stabilized after it is enabled by default. 

Reviewer: Are un-emulatable features intended as a stop-gap during beta or is this intended to be a permanent property of the emulation versions feature?

If this is Beta only, let's add a line item in the graduation criteria about resolving this before GA.

Author: My intention is to have it as a permanent property of the emulation versions feature. The API forward compatibility is a workaround for non-emulatable controller features. But there could be really difficult cases where non of the options work. We should have something to handle that instead of letting the feature be emulated with unpredictable results.

Reviewer: I would state this very clearly in this section.  Something like "We intent to allow non-emulatable features even in GA because ..."

Author: outdated.

```

#### Non-Emulatable Features
In very rare cases, if the feature implementation history could not be preserved in the code base with reasonable maintenance cost, a feature could opt out of the compatibility version through the `NonEmulatableFeatures` list. 

> With the `NonEmulatableFeatures` list, the server would fail to start if

Reviewer: I would move this section below the "So this exception should..." section.  The first paragraph is alarming and following it up promptly with the limits may help manage the blood pressure of readers.

Author: outdated.

1. the `EmulationVersion != BinaryVersion` and

> 1. any feature in the `NonEmulatableFeatures` list is disabled by default at the emulation version and explicitly enabled by the `--feature-gates` flag.

Reviewer: What alternatives were considered here?  It's not clear to me WHY we're deciding to refuse to start apiservers that have a non-emulatable beta on vs. allowing the beta to be turned on but not guaranteeing emulation for the feature.

Having the apiserver refuse to start is a potential footgun for anyone with automation that uses emulation version, and also offers the capability to turn on beta features/APIs.

Author: Giving more thoughts to it, I think the alternative is the better way to go. Updated the section with pros and cons.

So this exception should only be used if the feature  
1. is Beta and disabled by default for at least one version in the range of `1.{binaryMinorVersion-3}..1.{binaryMinorVersion-1}`.
1. is going through a lot of changes when it is off by default, and the feature implementation history is very hard to preserve.
1. has stabilized after it is enabled by default. 
1. Removing the feature out of `NonEmulatableFeatures` should be part of the feature's GA criteria to ensure the list would not just keep growing.

> An alternative for this option is just emulating these changing features at best effort, and make sure the clients are aware of the risks of enabling default-off Beta features with emulation version.

Reviewer: Ah, this was the alternative I expected to see from the above comment.  Can we offer more rationale here WHY we picked the proposed design over this alternative (pros and cons seem appropriate). 

At a glance this alternative looks better to me, so I'm curious to see the details of the decision.

Author: outdated.

```

#### Non-Emulatable Features
Generally, if a feature is enabled by default either in Beta or GA, it should be fairly stable already.
In very rare cases for some default-off Beta features, a feature could be non-emulatable if the feature implementation history could not be preserved in the code base with reasonable maintenance cost.

For these cases, we should try to emulate these changing features at best effort, and make sure the clients are aware of the risks of enabling default-off Beta features with emulation version.

##### Alternatives

> An alternative would be to keep a exception list of `NonEmulatableFeatures`, and make the server fail to start if any of the `NonEmulatableFeatures` are enabled in compatibility mode. This approach has the benefits of predictable outcomes. The downsides are:

Reviewer 2: I think we already have this exception list, all the feature gates that are beta and disabled by default are `NonEmulatableFeatures`

Author: do you mean the users are aware of the risks if they explicitly enable the disabled-by-default beta features?

1. whether a feature should be added to the list is subjective. It is hard to define a threshold to decide if the difficulty to preserve the implementation history is above this threshold, we could add the feature to the list.
1. this could break automated upgrade processes using the emulation version, depending what features a cluster opts in.

> 1. there needs proper cleaning up rules to ensure the list would not just keep growing.

Reviewer 2: FYI Jordan and David did something similar for avoiding beta apis to be enabled by default https://github.com/kubernetes/kubernetes/commit/af99d192cf95cd29e455d186878db409b8cbdf2c

### API availability

> Ideally, similar to feature flags, all APIs group-versions declarations should be modified

Reviewer 3: Is this referring to the prerelease lifecycle files or a different mechanism? prerelease lifecycle operates on the types instead of the group version.

Author: it is referring to the the prerelease lifecycle.

to track which Kubernetes version the API group-versions are introduced (or

> removed) at. But in practice, that would make code changes of API graduation intractable. 

Reviewer 3: intractable only for withholding API introductions. API removals don't need special handling.

Author: Updated the whole section. Please review again see if it is more clear.

> For example, if a controller is switching from Beta to GA API at the binary version, a full emulation would mean the whole controller code needs to be

Reviewer: My understanding is that we're doing this specifically for controllers.  Are there other use cases or should we simplify the wording here and just state up front that we're doing this to support controllers specifically?

duplicated to select from the Beta or GA API based on the emulation version every time the API is used. That means a lot of extra work, complicated testing rules, and high maintenance cost even for simple API graduations.

> Instead, we are proposing to keep forward compatibility of all AIPs in compatibility version mode: 

Reviewer 3: s/AIPs/APIs

> Instead, we are proposing to keep forward compatibility of all APIs in compatibility version mode: 

Reviewer: We're proposing that non-emulatable feature controllers to update to use the newest available API, ignoring emulation API enablement, right?  Maybe say this more directly?

Author: Yes, mainly for controllers for now. May expand to kublet if we want to add compat version in node. Updated the section.

All APIs group-versions declarations would be modified
to track which Kubernetes version the API group-versions are removed at. APIs introduced between the emulation version and 

> the binary version are still available if the stability level is no less than the APIs introduced before or at the emulation version and enabled at the emulation version.

Reviewer 3: Most of this is referring to API enablement/disablement on a group-version level. However, API lifecycles are generally individual on the group-version-resource level. I think it would make more sense to discuss the individual beta->GA resources rather than the entire group version?

Author: updated.

> the binary version are still available if the stability level is no less than the APIs of the same group and resource type introduced before or at the emulation version and enabled at the emulation version.

Reviewer: Could we get an example here?  This is twisty.

Author: Done.

> If an API has GAed and has not been removed at the emulated version, it would be enabled by default at the emulation version. If a newer stable version of the GA API has been introduced between the emulation version and the binary version, the new GA API would also be enabled at the emulation version along with the old GA API.

Reviewer: A comparison (maybe in a table), or how emulation works for normal APIs vs how it works for un-emulatable (non-emulatable?) features might make this information easier to digest.

If an API is Beta at the emulation version (meaning the Beta API has been introduced and has not been removed by the emulation version), it can be enabled by `--runtime-config=api/<version>` or it can be on-by-default at the emulation version. If a Beta API is enabled at the emulation version, and it has GAed between the emulation version and the binary version, its GA version would also be enabled at the emulation version.

Alpha APIs may not be enabled in conjunction with emulation version.

> The forward compatibility of APIs should not affect data compatibility because storage version is still controlled by the `MinCompatibilityVersion` regardless of whether the data are created through future versions of the API endpoints. Webhooks should also work fine if the matching policy is `Equivalent`.

Reviewer 3: I think we should add a section on how `MinCompatibilityVersion` will change storage version since the policy has already been n-1.

Author: Updated in https://github.com/kubernetes/enhancements/pull/5006

		cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc},
		func(options *metav1.ListOptions) {
			options.FieldSelector = fields.OneTermEqualSelector("metadata.name", DefaultServiceCIDRName).String()
		})

-	c.serviceCIDRLister = networkingv1beta1listers.NewServiceCIDRLister(c.serviceCIDRInformer.GetIndexer())
+	c.serviceCIDRLister = networkingv1listers.NewServiceCIDRLister(c.serviceCIDRInformer.GetIndexer())
	c.serviceCIDRsSynced = c.serviceCIDRInformer.HasSynced

	return c
  //...

type Controller struct {
	eventRecorder    record.EventRecorder

	serviceCIDRInformer cache.SharedIndexInformer
-	serviceCIDRLister   networkingv1beta1listers.ServiceCIDRLister
+	serviceCIDRLister   networkingv1listers.ServiceCIDRLister
	serviceCIDRsSynced  cache.InformerSynced

```
To fully emulate the controller for an older version, anywhere v1 api/type is referenced, it would need to switch to the v1beta version if the emulation version is older than the binary version. This would mean a lot of extra work, complicated testing rules, and high maintenance cost even for simple API graduations, while the emulation fidelity is unreliable with the extra complexity.

> So instead of truly emulating the feature controllers and API availability at the emulation version, we are proposing to keep forward compatibility of all APIs in compatibility version mode and have the non-emulatable feature controllers to update to use the newest available API. 

Reviewer 2: Let's say we add or remove field when we graduate from v1beta1 in 1.32 to v1 in 1.33, we may not be able to emulate in the 1.33 binary the behavior of the 1.32 controller  ...

Reviewer 4: If we add a v1 in 1.33, we still need to support clients using v1beta1. So we still should have the code and thus be able to emulate a "v1beta1 only" mode of operation.

One implication of this KEP is probably that we need to keep supporting older versions and behavior for longer than before, but I haven't looked into the details of that yet and could be wrong.

Reviewer 5: So if a 1.32 controller does thing/A.  A 1.33 controller does thing/A and also does wrongThing/B. Does this meant that switching to an emulation version for 1.32 does not stop the 1.33 controller from doing wrongThing/B?

Reviewer 2: one thing is to support fields and other support the behaviors associated to those fields, what I'm trying to point out is that the controller in 1.33 has a different behavior than  the one 1.32, that will not now to use this new field ... however, this is solved with this https://docs.google.com/document/d/1eLulBZ988bWNwrwnz9V70xVbQmHOYjrQpXRwiGmgcFM/edit?disco=AAABYEhCDY0 , if beta to GA does not have change on behavior

For API availability, this means:
1. if an API is removed (as indicated by the GVK prerelease lifecycle) after the emulation version, it would still be available at the emulation version. 
1. if an API is Beta at the emulation version (meaning the Beta API has been introduced and has not been removed by the emulation version), it can be enabled by `--runtime-config=api/<version>` or it can be on-by-default at the emulation version. If a Beta API is enabled at the emulation version, and it has GAed between the emulation version and the binary version, its GA version would also be enabled at the emulation version. if a newer Beta API is introduced between the emulation version and the binary version, the newer Beta API would also be enabled at the emulation version.

> 1. If an API has GAed and has not been removed at the emulated version, it would be enabled by default at the emulation version. If a newer stable version of the GA API has been introduced between the emulation version and the binary version, the new GA API would also be enabled at the emulation version along with the old GA API.

Reviewer 2: I think the main goal of this feature is to avoid to create new API objects in the storage that put us on a state we can not roll back, so this will require for this GA APIs to be stored in the API version of the emulation (I think we are alreadydoing that today)

Reviewer 2: I see, it is mentioned below,

1. Alpha APIs may not be enabled in conjunction with emulation version.

Here are some examples for `BinaryVersion = 1.33`:
API Prerelease Lifecycle | EmulationVersion | APIs Available @EmulationVersion 
-----|-----|-----
`v1alpha1: introduced=1.30, removed=1.31`<br>`v1beta1: introduced=1.31, removed=1.32`<br>`v1: introduced=1.32` | 1.30 | NA because we do not support emulating alpha apis.
`v1alpha1: introduced=1.30, removed=1.31`<br>`v1beta1: introduced=1.31, removed=1.32`<br>`v1: introduced=1.32` | 1.31 | `api/v1beta1` and `api/v1` available only when `--runtime-config=api/v1beta1=true`

> `v1alpha1: introduced=1.30, removed=1.31`<br>`v1beta1: introduced=1.31, removed=1.32`<br>`v1: introduced=1.33` | 1.33 | `api/v1`

Reviewer 6: There appears to be a gap here: beta removed in 1.32, but v1 introduced in 1.33.

```suggestion
`v1alpha1: introduced=1.30, removed=1.31`<br>`v1beta1: introduced=1.31, removed=1.32`<br>`v1: introduced=1.32` | 1.33 | `api/v1`
```

Author: ah, good catch. updated. 
Thank you!

> `v1beta1: introduced=1.31, removed=1.32`<br>`v1beta2: introduced=1.32` | 1.31 | `api/v1beta1` and `api/v1beta2` available only when `--runtime-config=api/v1beta1=true`

Reviewer 5: why is v1beta2 on in this scenario since it wasn't even introduced in 1.31?

Author: v1beta2 and v1beta1 have the same stability level, so the controller may have moved on to v1beta2 in the binary. By forward compatibility, we are exposing apis that are introduced later than the emulation version that are the same / more stable then the apis enabled at the emulation version.

Reviewer 5: Newer APIs can have new behavior that may be problematic and be the reason someone is trying to rollback.  For instance, what if v1beta2 broke cost estimation for VAP and someone wants to close the hole.  They rollback to 1.31 behavior, but because this API is being enabled when only v1beta1 is wanted, they're stuck with the improper version enabled.  That seems like the sort of scenario we want to avoid.

Reviewer 5: I bet this is stemming from different modes of upgrade.  Some systems upgrade all the kube binaries as a unit (I think GKE does this) and while they may rollout in an HA manner, each unit is either 1.31 or 1.32.  Other system upgrade binaries across the cluster, so each instance of the kube-apiserver upgrades, then the kube-controller-manager, etc.  OpenShift does this.

This leads to two different perspectives of emulation version. 
1. I want `kube-apsierver --emulation-version=1.31` to be explicitly different than a 1.31 kube-apiserver.  Perhaps in a manner that makes it impractical to stop taking broken paths (v1beta2 is broken case).  This might be preferred (sometimes) by upgrade everything as a unit upgrades. But leaves us in a state where if v1beta2 is broken, there doesn't appear to be a way to actually disable it and function.
2. I want `kube-apiserver --emulation-version=1.31` to be behave like a 1.31 kube-apiserver.  This is probably preferred by those who upgrade binaries independent of each other.

The current proposal appears to make #2 impossible.  What if instead we made `--emulation-version=1.31` emulate the 1.31 version of the kube-apiserver and required requesting serving the non-included APIs?  In this example `--runtime-config=api/v1beta1=true,api/v1beta2=true`

Author: To avoid the confusion about what is expected from emulation version, I have added a `--api-forward-compatible`, so that `kube-apiserver --emulation-version=1.31` would behave like a 1.31 kube-apiserver by default, and only include the future api if explicitly set.

Reviewer: I think I prefer `--runtime-config=api/v1beta1=true,api/v1beta2=true` over  `--api-forward-compatible`.

This way the command line flag has to explicitly opt-in to enabling the APIs that will be enabled.

Either way we need to decide what happens if the controller has been authored to reconcile the v1beta2 API but emulation version is set to the older minor version and `--runtime-config=api/v1beta1=true`.  Clearly the controller will not work.  I think I'd prefer we detect that and disallow it up-front during flag parsing

Reviewer 5: `--api-forward-compatible` would need a version correct?  1.32 and 1.33 and 1.34 could all have different APIs they need served.

Once benefit to `--api-forward-compatible` is that cluster-admins don't have to know which APIs they need to enable, which could be pretty handy.  Particularly over longer skews that will need to enable v1 APIs.

Reviewer: > `--api-forward-compatible` would need a version correct? 1.32 and 1.33 and 1.34 could all have different APIs they need served.

I like this idea.  I'm struggling a bit to convince myself that we're going to have something that doesn't break controllers.  If our motivation for all of this is to allow a beta controller author to write the controller against the latest available API and know it works even when emulation version is enabled, then we're locked into what API versions need to be enabled.  This probably works is we require that the flag be set to a known good value.  Spelling out the allowed combinations of this flag and `--emulate-version` in such a way that we can't get into broken states would satisfy me.

> One benefit to `--api-forward-compatible` is that cluster-admins don't have to know which APIs they need to enable, which could be pretty handy. Particularly over longer skews that will need to enable v1 APIs.

Agreed.  I think I'm down to nits about the flag name.  Are we intending for this to solely enable APIs needed by beta API controllers?  If so, should we name it something more targeted like --emulate-version-forward-compatible` ?

Reviewer 5: >  If our motivation for all of this is to allow a beta controller author to write the controller against the latest available API 

I think it's more than just the beta levels.  For instance, across three versions we have a state like this
* 1.31, controller/x uses v1beta1
* 1.32, controller/x uses v1beta2
* 1.33, controller/x uses v1

customer/A and customer/B at 1.31, we the user ran with `--runtime-config=v1beta1=true` an thus controller/x ran.  

customer/A is now at 1.33, let's say they did this an offline upgrade with everything off.  Such an upgrade story can upgrade over any number of versions at once.  Now there's a problem and they want to go back to `--emulation-version=1.31 --emulate-version-forward-compatible=1.33 --runtime-config=v1beta1=true`.  In this situation they need the v1 API enabled also.

customer/B is now at 1.32.  There's a problem and they want to back to  `--emulation-version=1.31 --emulate-version-forward-compatible=1.32 --runtime-config=v1beta1=true`.  In this situation they need the v1beta2 API enabled also.

That seems pretty user friendly because the directions are simply to add `--emulation-version=1.31 emulate-version-forward-compatible=1.32` to their old launch params.

If we had customer/C, who ran 1.31 without v1beta1=true. They end up at 1.33.  They want to go back to `--emulation-version=1.31 --emulate-version-forward-compatible=1.33`.  I would expect this to not enable the v1 API and I would expect controller/x to not run.  Is this what would happen?

Reviewer: > in this situation they need the v1 API enabled also.

This is exactly the type of scenario I'd like to make sure we support correctly.  This scenario works when `-emulate-version-forward-compatible=1.33` since the binary version is 1.33 and the 1.33 provides v1 of the API, which is what the controller expects.  I wouldn't know what `--emulate-version-forward-compatible=1.32` would mean for 1.31 emulation with a 1.33 binary.  Does this suggest that `--emulate-version-forward-compatible` should always be set to the binary version?

> If we had customer/C, who ran 1.31 without v1beta1=true. They end up at 1.33. They want to go back to `--emulation-version=1.31 --emulate-version-forward-compatible=1.33`. I would expect this to not enable the v1 API and I would expect controller/x to not run. Is this what would happen?

That's my expectation as well.  I'm expecting the apiserver to only enable forward compatible APIs for the APIs that are actually enabled, and that we've flagged as needing forward compatibility.

Reviewer 3: > This is exactly the type of scenario I'd like to make sure we support correctly. This scenario works when -emulate-version-forward-compatible=1.33 since the binary version is 1.33 and the 1.33 provides v1 of the API, which is what the controller expects. 

I think we're missing a case in the KEP of this scenario with the chain of `v1beta1->v1beta2->v1`. Logically (at`--emulated-version=1.31 --forward-compatible=1.33` bv: 1.33) enabling v1beta1 would only require v1 to be enabled (for controller compatibility), but what about other APIs in the chain? Would v1beta2 be need to be enabled as well? Is it weird that upgrading from `--emulated-version=1.31 --forward-compatible=1.32` (bv=1.32) to `--emulated-version=1.31 --forward-compatible=1.33` (bv=1.33) disables the v1beta2 API?

> I wouldn't know what --emulate-version-forward-compatible=1.32 would mean for 1.31 emulation with a 1.33 binary. Does this suggest that --emulate-version-forward-compatible should always be set to the binary version?

I would think so, we cannot guarantee forward compatibility with another version other than the binary version.

Author: suppose the controller uses `v1beta1 (bv: 1.31)`-> `v1beta2 (bv: 1.32)` -> `v1 (bv: 1.33)`.
If the user is using `--emulated-version=1.31 (bv: 1.32)` to make the controller work, we have to enable `v1beta2` when `--runtime-config=v1beta1=true`. If the user is using `--emulated-version=1.31 (bv: 1.33)` to make the controller work, we have to enable `v1` when `--runtime-config=v1beta1=true`, in this case setting `--emulate-version-forward-compatible=1.32` does not make the controller work with `v1beta2` api because the controller is already using the lastest API. So we have to enable the latest api. 
As for whether `v1beta2` should be enabled: from the api lifecycle, we know that `v1beta2` is introduced at 1.32 and `v1` at 1.33. But we don't know when the controller is changed from `v1beta2` to `v1`, because sometimes an api is introduced before it is actually used. From the perspective of the api server, I think we have enable both `v1beta2` and `v1` when `--emulate-version-forward-compatible` is requested.

Reviewer: It sounds like we've mostly converged on the problem and what the general shape of a solution should be?  How about a KEP update that incorporates all of this?  We could then iterate from there.

Author: Updated. PTAL

`v1beta1: introduced=1.31, removed=1.32`<br>`v1beta2: introduced=1.32` | 1.33 | `api/v1beta2` available only when `--runtime-config=api/v1beta2=true`
`v1: introduced=1.28, removed=1.32`<br>`v2beta1: introduced=1.31, removed=1.32`<br>`v2: introduced=1.32` | 1.30 | `api/v1`
`v1: introduced=1.28, removed=1.32`<br>`v2beta1: introduced=1.31, removed=1.32`<br>`v2: introduced=1.32` | 1.31 | `api/v1` always available, `api/v2beta1` and `api/v2` available only when `--runtime-config=api/v2beta1=true`

> `v1: introduced=1.28, removed=1.32`<br>`v2beta1: introduced=1.31, removed=1.32`<br>`v2: introduced=1.32` | 1.33 | `api/v2`

Reviewer 2: we don't remove v1 apis, https://kubernetes.io/docs/reference/using-api/deprecation-policy/#deprecating-parts-of-the-api

> Rule #4a: API lifetime is determined by the API stability level
> GA API versions may be marked as deprecated, but must not be removed within a major version of Kubernetes

For the controller, at the emulation version the controller is still enabled by enabling the Beta API **AND** the controller feature as before, but under the hood the controller is calling the newer API. 

The forward compatibility of API availability should not affect data compatibility because storage version is still controlled by the `MinCompatibilityVersion` regardless of whether the data are created through future versions of the API endpoints. Webhooks should also work fine if the matching policy is `Equivalent`.

> For the use case of upgrading control plane binary version without changing the emulation version, this would mean api servers should be upgraded first before any other components, because an api server of the old binary version would not be able to serve a controller of the new binary version even though the emulation version is the same as the old binary version.

Reviewer 2: is not the case today? apiserver should be always be upgraded first

Author: today although apiserver should be upgraded first according to the doc, you can still upgrade all the control plane components in one node first - apiserver1, cm1, scheduler1, before moving on to the next node. That is not a problem today because apis support N-1, N, N+1. But with emulation version, this might not work anymore.

Reviewer 3: Coordinated Leader Election is trying to solve this problem but it is preexisting. Given the situation Siyuan described, a controller that graduated to GA could still talk to a GA API unavailable on other apiservers that haven't upgraded.

#### Alternatives to API forward compatibility

> To make API graduation workable for controller code change under compatibility version, we have considered and rejected the following alternative options:

Reviewer 2: This also always assume that the logic in the controllers is the same between versions, but that may not be the case, if the controller and the API change between versions, or also when an existing GA API add a new field and the controller depends on that new field, you'll never being able to emulate the previous behavior ...

Author: similar to https://github.com/kubernetes/enhancements/pull/5038#discussion_r1924180177

### API availability

Ideally, similar to feature flags, all API's group-version-resource declarations should be modified
to track which Kubernetes version the GVRs are introduced (or
removed) at: if an API is introduced after (or removed before) the emulation version, it should not be available at the emulation version. 

> But in practice, that would make the controller code changes intractable if an API is graduating from Beta to GA and the controller wants to use newer API. For example, to graduate Multiple Service CIDRs to GA, normally the controller code change would look like:

Reviewer: ```suggestion
But in practice, that would make the controller code changes unreasonably burdonsome if an API is graduating from Beta to GA and the controller wants to use newer API. For example, to graduate Multiple Service CIDRs to GA, normally the controller code change would look like:
```

(It's pretty clear to me that this is possible, I just don't feel comfortable asking people to carry the burden)

	serviceCIDRsSynced  cache.InformerSynced

```
To fully emulate the controller for an older version, anywhere v1 api/type is referenced, it would need to switch to the v1beta version if the emulation version is older than the binary version. This would mean a lot of extra work, complicated testing rules, and high maintenance cost even for simple API graduations, while the emulation fidelity is unreliable with the extra complexity.

So instead of truly emulating the feature controllers and API availability at the emulation version, we are allowing apis introduced after the emulation version to be enabled explicitly with `--runtime-config=api/<version>`. We are also introducing an umbrella `--emulation-forward-compatible` flag to enable forward compatibility of all APIs in compatibility version mode: if an older version of the API is enabled, newer and more stable versions of the same resource group introduced between the emulation version and binary version would also be enabled. This way the non-emulatable feature controllers would be able to use the newest available API. 

For API availability, setting `--emulation-forward-compatible=true` means:

> 1. if an API is removed (as indicated by the GVK prerelease lifecycle) after the emulation version, it would still be available at the emulation version. 

Reviewer 7: _Removed_ APIs still being available when emulating old versions is driven solely by `--emulation-version`, regardless of `--emulation-forward-compatible=true`, right?

Author: yes, that's correct. APIs removed after the `--emulation-version` will still be available regardless of `--emulation-forward-compatible=true`.

> 1. if an API is Beta at the emulation version (meaning the Beta API has been introduced and has not been removed by the emulation version), it can be enabled by `--runtime-config=api/<version>` or it can be on-by-default at the emulation version. If a Beta API is enabled at the emulation version, and it has GAed between the emulation version and the binary version, its GA version would also be enabled at the emulation version. If a newer Beta API is introduced between the emulation version and the binary version, the newer Beta API would also be enabled at the emulation version.

Reviewer 7: nit: there could be multiple GA or newer beta API versions

```suggestion
1. if an API is Beta at the emulation version (meaning the Beta API has been introduced and has not been removed by the emulation version), it can be enabled by `--runtime-config=api/<version>` or it can be on-by-default at the emulation version. If a Beta API is enabled at the emulation version, and it has GAed between the emulation version and the binary version, its GA version(s) would also be enabled at the emulation version. If a newer Beta API is introduced between the emulation version and the binary version, the newer Beta API(s) would also be enabled at the emulation version.
```

1. if a Beta API is not enabled at the emulation version, its future versions would not be enabled at the emulation version unless explicitly enabled with `--runtime-config=api/<version>`.

> 1. If an API has GAed at the emulated version, it would be enabled by default at the emulation version. If a newer stable version of the GA API has been introduced between the emulation version and the binary version, the new GA API would also be enabled at the emulation version along with the old GA API.

Reviewer 7: ```suggestion
1. If an API has GAed at the emulated version, it would be enabled by default at the emulation version. If a newer stable version of the GA API has been introduced between the emulation version and the binary version, the new GA API(s) would also be enabled at the emulation version along with the old GA API.
```

1. Alpha APIs may not be enabled in conjunction with emulation version.

Here are some examples for `BinaryVersion = 1.33`:
API Prerelease Lifecycle | EmulationVersion | APIs Available @EmulationVersion 
-----|-----|-----
`v1alpha1: introduced=1.30, removed=1.31`<br>`v1beta1: introduced=1.31, removed=1.32`<br>`v1: introduced=1.32` | 1.30 | NA because we do not support emulating alpha apis.
`v1alpha1: introduced=1.30, removed=1.31`<br>`v1beta1: introduced=1.31, removed=1.32`<br>`v1: introduced=1.32` | 1.31 | `api/v1beta1` available when `--runtime-config=api/v1beta1=true`<br>`api/v1beta1` and `api/v1` available only when `--runtime-config=api/v1beta1=true,api/v1=true` or `--runtime-config=api/v1beta1=true --emulation-forward-compatible=true`
`v1alpha1: introduced=1.30, removed=1.31`<br>`v1beta1: introduced=1.31, removed=1.32`<br>`v1: introduced=1.32` | 1.33 | `api/v1`

> `v1beta1: introduced=1.31, removed=1.32`<br>`v1beta2: introduced=1.32` | 1.31 | `api/v1beta1` available when `--runtime-config=api/v1beta1=true`<br>`api/v1beta1` and `api/v1beta2` available only when `--runtime-config=api/v1beta1=true,api/v1beta2=true` or `--runtime-config=api/v1beta1=true --emulation-forward-compatible=true`

Reviewer 5: This matches expectations. Thank you for the update.

Beta feature graduated to GA|on-by-default|on|feature enabled unless `--feature-gates=<feature>=false`|feature always enabled, feature gate may not be set
Beta feature removed|on-by-default|-|feature enabled unless `--feature-gates=<feature>=false`|feature does not exist
Alpha API introduced|-|off-by-default|API does not exist|alpha APIs may not be used in conjunction with emulation version

> Beta API graduated to GA|off-by-default|on|API `api/v1beta1` and `api/v1` available only when `--runtime-config=api/v1beta1=true`|API `api/v1` available

Reviewer 6: Does "graduated to GA" imply that the Beta API is removed (from N)?

Should `--runtime-config=api/v1beta1=true` have some effect when **emulation-version=N**?

Author: "graduated to GA" does not imply that the Beta API is removed (from N) automatically. 
`--runtime-config=api/v1beta1=true` would have both `api/v1beta1` and `api/v1` both available. This is the same as before, so we are not listing it out explicitly.

Beta API removed|off-by-default|-|API `api/v1beta1` available only when `--runtime-config=api/v1beta1=true`|API `api/v1beta1` does not exist
on-by-default Beta API removed|on-by-default|-|API `api/v1beta1` available unless `--runtime-config=api/v1beta1=false`|API `api/v1beta1` does not exist

> API Storage version changed|v1beta1|v1|Resources stored as v1beta1|Resources stored as v1

Reviewer 6: Does the storage version depend on the compatibility version? If so, should it be mentioned here?

What happens in this case for **Story 2**, administrator needs to roll back?

1. 1pm: emulation=N-1, everything is stored v1beta1
2. 2pm: emulation=N, so some items are stored as v1
3. 3pm: emulation=N-1, so...?
